stateresolution.go

```diff
@@ -0,0 +1,255 @@
+// and works out which event should be used for each state event.
+func ResolveStateConflicts(conflicted []Event, authEvents []Event) []Event {
+	var r stateResolver
+	// Group the conflicted events by type and state key.
+	r.addConflicted(conflicted)
+	// Add the unconflicted auth events needed for auth checks.
+	for i := range authEvents {
+		r.addAuthEvent(&authEvents[i])
+	}
+	// Resolve the conflicted auth events.
+	r.resolveAndAddAuthBlocks([][]Event{r.creates})
+	r.resolveAndAddAuthBlocks([][]Event{r.powerLevels})
+	r.resolveAndAddAuthBlocks([][]Event{r.joinRules})
+	r.resolveAndAddAuthBlocks(r.thirdPartyInvites)
+	r.resolveAndAddAuthBlocks(r.members)
+	// Resolve any other conflicted state events.
+	for _, block := range r.others {
+		if event := r.resolveNormalBlock(block); event != nil {
+			r.result = append(r.result, *event)
+		}
+	}
+	return r.result
+}
+
+// A stateResolver tracks the internal state of the state resolution algorithm
+// It has 3 sections:
+//
+//  * Lists of lists of events to resolve grouped by event type and state key.
+//  * The resolved auth events grouped by type and state key.
+//  * A List of resolved events.
+//
+// It implements the AuthEvents interface and can be used for running auth checks.
+type stateResolver struct {
+	creates                   []Event
+	powerLevels               []Event
+	joinRules                 []Event
+	thirdPartyInvites         [][]Event
+	members                   [][]Event
+	others                    [][]Event
+	resolvedCreate            *Event
+	resolvedPowerLevels       *Event
+	resolvedJoinRules         *Event
+	resolvedThirdPartyInvites map[string]*Event
+	resolvedMembers           map[string]*Event
+	result                    []Event
+}
+
+func (r *stateResolver) Create() (*Event, error) {
+	return r.resolvedCreate, nil
+}
+
+func (r *stateResolver) PowerLevels() (*Event, error) {
+	return r.resolvedPowerLevels, nil
+}
+
+func (r *stateResolver) JoinRules() (*Event, error) {
+	return r.resolvedJoinRules, nil
+}
+
+func (r *stateResolver) ThirdPartyInvite(key string) (*Event, error) {
+	return r.resolvedThirdPartyInvites[key], nil
+}
+
+func (r *stateResolver) Member(key string) (*Event, error) {
+	return r.resolvedMembers[key], nil
+}
+
+func (r *stateResolver) addConflicted(events []Event) {
+	type conflictKey struct {
+		eventType string
+		stateKey  string
+	}
+	offsets := map[conflictKey]int{}
+	// Split up the conflicted events into blocks with the same state key.
+	for _, event := range events {
+		key := conflictKey{event.Type(), *event.StateKey()}
+		// Work out which block to add the event to.
+		var block *[]Event
+		// By default we add the event to a block in the others list.
+		blockList := &r.others
+		switch key.eventType {
+		case "m.room.create":
+			if key.stateKey == "" {
+				block = &r.creates
+			}
+		case "m.room.power_levels":
+			if key.stateKey == "" {
+				block = &r.powerLevels
```

Comment:
Right, I think I know what you're doing, and I think this function is needlessly complicated considering what it's doing. I think it's confusing because you're desperately trying to only have 1 place where you append to the list, so you go to great lengths to end up assigning `block` to the right `[]Event` which obfuscates what you're actually doing, which really does appear to be just grouping events into pre-defined lists based on the `conflictKey`.

Look at `:108` for a good example of why I think this is confusing. If `block == nil` it also implies that you need to look at `blockList`, but you can't do `if blockList != nil` because you have already defaulted it to the `others` list.

The entire point is to "do checks, assign to list, continue" and I would prefer if the code reflected that better e.g:

```go
for _, event := range events {
    if event.Type() == "m.room.create" && *event.StateKey() == "" {
        r.creates = append(r.creates, event)
        continue
    }
    // etc...
}
```

I don't understand why you store events where there's custom state keys (e.g member events) as a 2d array rather than a map which feels more natural. Why do you do that? If you didn't, you wouldn't need to maintain this `offsets` structure.